src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs

```diff
@@ -69,100 +68,72 @@ internal partial class SQLitePersistentStorage
-            // writers there are, and which only lets us past once all the concurrent writers
-            // say they are done.
-            CountdownEvent countdown;
-
-            // Note: by blocking on _writeQueueGate we are guaranteed to see all the writes 
-            // performed by FlushAllPendingWrites.
-            using (await _writeQueueGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
-            {
-                // Get the writes we need to process.
-                writesToProcess.AddRange(keyToWriteActions[key]);
+            // Get the task that is responsible for doing the writes for this queue.
+            // This task will complete when all previously enqueued writes for this queue
+            // complete, and all the currently enqueued writes for this queue complete as well.
+            var writeTask = await GetWriteTaskAsync().ConfigureAwait(false);
+            await writeTask.ConfigureAwait(false);
 
-                // and clear them from the queues so we don't process things multiple times.
-                keyToWriteActions.Remove(key);
+            return;
 
-                // We may have acquired _writeQueueGate between the time that an existing thread 
-                // completes the "Wait" below and grabs this lock.  If that's the case, let go
-                // of the countdown associated with this key as it is no longer usable.
-                RemoveCountdownIfComplete(keyToCountdown, key);
-
-                // See if there's an existing countdown keeping track of the number of writers
-                // writing this queue.
-                if (!keyToCountdown.TryGetValue(key, out countdown))
-                {
-                    // We're the first writer for this queue.  Set the count to one, and keep
-                    // it around so future concurrent writers will see it.
-                    countdown = new CountdownEvent(initialCount: 1);
-                    keyToCountdown.Add(key, countdown);
-                }
-                else
-                {
-                    // If there is, increment the count to indicate that we're writing as well.
-                    countdown.AddCount();
-                }
-
-                Debug.Assert(countdown.CurrentCount >= 1);
-            }
-
-            // Now actually process any writes we found for this queue.
-            ProcessWriteQueue(connection, writesToProcess);
-
-            // Mark that we're done writing out this queue, and wait until all other writers
-            // for this queue are done.  Note: this needs to happen in the lock so that 
-            // changes to the countdown value are observed consistently across all threads.
-            bool lastSignal;
-            using (await _writeQueueGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            // Local functions
+            async Task<Task> GetWriteTaskAsync()
```

Comment:
💭 Would prefer to see this as a separate method instead of a local function due to its size.